source/Optimizers/nelder_mead.d

```diff
@@ -0,0 +1,519 @@
+        Vector!float[] y_train = new Vector!float[num_points];
+        Vector!float[] y_tilde = new Vector!float[num_points];
+
+        foreach(i; 0 .. num_points) {
+            x_train[i] = point_training;
+            y_train[i] = new Vector!float(len);
+            y_tilde[i] = new Vector!float(len);
+            y_train[i].v[] = x_train[i].v[];
+        }
+
+        // Create Neural Network.
+        auto nn = new NeuralNetwork!float(len);
+        nn.func!("b.v[] = p0.v[];", Vector!float)
+                (2, null, null, null, null, [2], [0.0])
+          .serialize;
+
+        float loss_function_linRel(in Vector!float _v) {
+            float loss_value = 0.0;
+
+            // We equipe the neural network with the weigth given in parameters.
+            nn.set_parameters(_v);
+
+            // We loop over all data points and compute the sum of squared errors.
+            foreach(i; 0 .. num_points) {
+                nn.apply(x_train[i], y_tilde[i]);
+                y_tilde[i] -= y_train[i];
+                loss_value += y_tilde[i].norm!"L2";
+            }
+
+            return loss_value/num_points;
+        }
+
+        auto sol = new Vector!float(nn.serialized_data.length, 0.0);
+        auto res = nelder_mead!float(sol, &loss_function_linRel, 1.0, 1);
+
+        sol -= expected_solution;
+        assert (sol.norm!"L2" < 1e-3, "Optimizers: Nelder_Mead: " ~ str_action);
+    }
+
+    void test_action_NM_shrink(Vector!float expected_solution,
+                               string str_action,)
+    {
+        float loss_function_cosabs(in Vector!float _v) {
+            auto v0 = 0.5*(_v[1] - _v[0]);
+            auto v1 = 0.5*(1+sqrt(2.0))*(_v[0]+_v[1]-1.0);
+            auto z = sqrt(v0*v0 + v1*v1);
+            return z - cos(2.0*3.1415926535*z);
+        }
+
+        auto sol = new Vector!float(2, 0.0);
+        auto res = nelder_mead!float(sol, &loss_function_cosabs, 1.0, 1);
+
+        sol -= expected_solution;
+        assert (sol.norm!"L2" < 1e-3, "Optimizers: Nelder_Mead: " ~ str_action);
+    }
+
+    auto vec_r1 = new Vector!float([0, 1]);
+    auto vec_r2 = new Vector!float([1.0+1.0/sqrt(2.0), 1.0+1.0/sqrt(2.0)]);
+    auto vec_e = new Vector!float([1.5+sqrt(2.0), 1.5+sqrt(2.0)]);
+    auto vec_c_i = new Vector!float([0.25-1.0/(2.0*sqrt(2.0)),
+                                     0.25-1.0/(2.0*sqrt(2.0))]);
+    auto vec_c_e = new Vector!float([0.75 + 1.0/(2.0*sqrt(2.0)),
+                                     0.75 + 1.0/(2.0*sqrt(2.0))]);
+    auto vec_s = new Vector!float([0.5, 0.5]);
+
+    test_action_NM(vec_r1, "Reflection - 1st case", new Vector!float([0.5, 1 + 1/sqrt(2.0)]));
+    test_action_NM(vec_r2, "Reflection - 2nd case");
+    test_action_NM(vec_e, "Expension");
+    test_action_NM(vec_c_i, "Contraction (Internal)", new Vector!float([0.1, 0.1]));
+    test_action_NM(vec_c_e, "Contraction (External)", new Vector!float([0.9, 0.9]));
+    test_action_NM_shrink(vec_s, "Shrink");
+    writeln("Done");

```

Comment:
How much time does this take start-to-finish?

Reply:
Less than a second 💃